### DISCUSSION TITLE
Refactoring the key handling

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3916] [Development]
This is **every** function involved in key handling.
On the far right is the curses function `getch()`, and every left-hand leaf is a consumer.

A few consumers, like `mutt_what_key()` use the keycode directly, but the majority expect an opcode, e.g. `OP_DELETE`.

<img src="https://flatcap.org/mutt/keys/g3-compact.svg">

**Source**: [graphviz](https://flatcap.org/mutt/keys/g3-compact.gv), [svg](https://flatcap.org/mutt/keys/g3-compact.svg)
**Key**:
- Red/Green/Blue: Functions of interest
- Yellow: Full Dialogs, see [Dialog Windows](https://neomutt.org/code/group__gui__dlg.html)
- Purple: Message Window functions, see [Message Windows](https://neomutt.org/code/group__gui__mw.html)
- Grey: Global data

---

`mutt_getch()` has three possible sources for input.
In order of priority:

1. `UngetKeyEvents`: a buffer of keys that have been pushed back.
   There are some functions that need to peek at the next character, but might not need it.
   They can `mutt_unget_ch()` to return it.

2. `MacroEvents`: a buffer of pending macro keystrokes.
   When a macro is encountered, it's keystrokes are pushed into this buffer, then `mutt_getch()` can be called as normal.
  This can be temporarily ignored by setting the global `OptIgnoreMacroEvents`.

3. `getch()`: an ncurses function that actually checks the keyboard.

---

`km_dokey_event()` is responsible for the key --\> opcode translations.
It's also responsible for calling `imap_keep_alive()` on timeouts.

`km_dokey_event()` takes `MenuType` parameter, e.g. `MENU_ALIAS`.
It uses the global `KeyMaps` to get a list of key bindings / macros (`struct Keymap`).

If it can't find a matching binding in the given menu, it tries the Generic bindings.
If they fail too, it searches **all** the menus for a match then ungets the function as a string `<name>`. 

`km_dokey_event()` includes business logic to provide special behaviour for the Editor, Pager and Generic menus.

**Comment by lucilanga** [comment]
Quoting from #3913 as it is directly related:

> Some final thoughts...
> 
> Each Dialog has set of mappings that are stored in `Keymaps[MENU_X]`. Most Dialogs also fallback to `Keymaps[MENU_GENERIC]`.
> 
> If we can pass multiple `Keymap` arrays into `km_dokey_event()`, then we can eliminate some of the logic (and `retry_generic()`).
> 
> The other `MENU_EDITOR` logic is to hide timeouts. It's convenient, but we can probably move that out to `mw_get_field()`

So I took km_dokey_event() -> km_dokey_event_line(), made it accept Keymap *map (I kept mtype for now). 
Took the map logic to an outside func for now: km_dokey_event2() where I'll handle the TAILQ, retry_generic() stuff, and push_string().
retry_generic() will be a call to km_dokey_event2() with MENU_GENERIC, I'll probably have to move mutt_unget_ch things from retry_generic to km_do_event_line(). Naming sucks for now but I'll clean this up later. imap carp is in limbo for now.

I know it's not much for now, but I need to know if I'm doing this right, or I'm completely off the course with this.
I'll have to  open a devel branch for this soon, so you can actually see anything.

PS. Most of the business logic went to km_dokey_event2()

**Reply by flatcap** [reply]
> So I took km_dokey_event() -> km_dokey_event_line()...
> I need to know if I'm doing this right

Yes, you sound like you understand things :-)

> Naming sucks for now

One of the hardest problems in Computer Science :-)

> imap carp is in limbo for now.

OK.

I think we could simplify things.
(Just ideas, I haven't checked if this'll work correctly)

Change to a fixed: `mutt_getch_timeout(1000)`

If it returns a real event, then handle it.

However, if it returns `OP_TIMEOUT`, then:

- Call `imap_keep_alive()`.
  It doesn't matter if we call it often, it already checks that enough time has elapsed.

- If `$timeout` seconds is reached, then return `OP_TIMEOUT` to the caller.
  A few callers, need `$timeout`, e.g. Index to handle `$new_mail_command`.

A better long-term solution might be:

- Create a new `struct Notify` object on `struct Neomutt` for timeouts

- Interested parties register an Observer, e.g.
  - IMAP - `imap_keep_alive()`
  - `$new_mail_command` -- currently handled by the Index and Pager

- `km_dokey_event()` sends a notification on timeouts

> I'll have to open a devel branch for this soon

I look forward to it.

**Comment by lucilanga** [comment]
Staring for too long at km_dokey_event() got me really confused.
I do understand what km_dokey_event currrently does, but I'm not sure I understand the problem we're trying to fix.

Is it related to the fact that we need to lookup several MENUS at once to try to find the correct opcodes.

Like what we're about to do for mw_get_field(),
we were looking at MENU_EDITOR, but we want to look at SIDEBAR and then fall to EDITOR.

Or is there a bigger picture in key opcode translation, that somehow I miss ?

**Reply by flatcap** [reply]
> Staring for too long at km_dokey_event() got me really confused.

Don't worry, NeoMutt's code has that effect on people :-)

> I do understand what km_dokey_event currrently does,
> but I'm not sure I understand the problem we're trying to fix.

First, we don't **need** to fix anything.

You added `(mtype == MENU_SIDEBAR)` changes to `km_dokey_event()`.
It's ugly, but it works.

You can stop reading here :-)

**Reply by flatcap** [reply]
However, it's a good opportunity to examine how these functions work and,
if you have the time and patience, to improve them.

> Is it related to the fact that we need to lookup several MENUS at once to try
> to find the correct opcodes.

Yes.

Some Dialogs rely on up to four different `*_function_dispatcher()`s

The Pager has:

- `pager_function_dispatcher()`
- `index_function_dispatcher()`
- `sb_function_dispatcher()`
- `global_function_dispatcher()`

These match the four "domains" of actions.
Those on the Pager itself, those on the parent Index, Sidebar functions and Global functions.

If you now look at `OpPager`, it lists **every** function across all four domains.
Ideally, we'd have `OpPager`, `OpIndex`, `OpSidebar`, `OpGlobal` each associated with a dispatcher function.

The first step is to change `km_dokey_event()` to take a set of `struct MenuFuncOp`.
If a caller needs Generic functions, then pass, say, `{ OpAlias, OpGeneric }`
This means we can eliminate the `(mtype == MENU_SIDEBAR)` logic and `retry_generic()`.

**Reply by flatcap** [reply]
> Or is there a bigger picture in key opcode translation, that somehow I miss ?

Looking **much** further ahead, I want to move the event loop **out** of the Dialog.

We'd need to split `OpPager` and friends into much smaller pieces.
Each would be associcated with a dispatcher function: `DomainPager = { OpPager, pager_function_dispatcher() }`

Some thought is needed on how keybindings will work.
Would we want `bind pager` to apply to all its domains,
or should we `bind pager` or `bind sidebar` or `bind global`, etc

A Dialog would define which domains it accepts: `DomainPager`, `DomainSidebar`, ...
Any setup / teardown could be moved into separate functions, leaving just the event loop.

The event loop of many Dialogs is quite small:

```c
do
{
  op = km_dokey(MENU_ALIAS);

  rc = alias_function_dispatcher(dlg, op);
  if (rc == FR_UNKNOWN)
    rc = menu_function_dispatcher(menu->win, op);
  if (rc == FR_UNKNOWN)
    rc = global_function_dispatcher(NULL, op);
} while (rc != FR_DONE);
```

If we've defined our domains, then this loop can be eliminated and replaced with a universal function.

Eventually, we'd have standardised Dialogs.
Each would have a constructor that defines its domains and initialises any data and a destructor to clean up afterwards.

The caller would:

```c
struct Dialog *dlg = alias_dialog_create(args);
int rc = dlg_run(dlg);
dlg_free(&dlg);
```

**Comment by lucilanga** [comment]
> The first step is to change `km_dokey_event()` to take a set of `struct MenuFuncOp`.
> If a caller needs Generic functions, then pass, say, `{ OpAlias, OpGeneric }`
> This means we can eliminate the `(mtype == MENU_SIDEBAR)` logic and `retry_generic()`.

So struct MenuFuncOp will act as a haystack, and probably KeyMap (MenuType right now) will act as the needle.
I'm perfectly fine with that and I'm sure km_dokey_event() will cleanup nicely.

However I'm unsure about cleaning up MenuFuncOp structures like OpAlias, OpPager, etc.
Maybe not a big problem but right now there is a somewhat of a duplication in MenuFuncOp and 
MenuOpSeq (aka DefaultBindings), well maybe not duplication, the content of MenuFuncOp and MenuOpSeq 
are somewhat in sync with each other.
Cleaning MenuFuncOp structures wrt _dispatcher() will probably create a mess wrt to code readability.
For instance I will have to track code all over the place to find how MENU_ALIAS is actually working:

eg:
km_dokey(MENU_ALIAS, {OpAlias, OpGeneric})
km_dokey(MENU_PAGER, {OpPager, OpIndex, OpSidebar, OpGeneric})

I'm sure this mess will cleanup in the 'much further ahead' phase where we will have DomainMENU like structures.

Come to think of it right now you we're saying the same thing in reply 2 and 3 (in reply 1 you're telling me to mind my own business - but I ignored that :) )
I suppose for now we can throw MenuFuncOp structures as they were and see where this takes us.

On the matter of keybindings in the later reply, keybindings is a matter that's bugging me for a while.
Probably because of early work on quick sidebar stuff, I have trouble of mentally isolating the behavior
of mw_get_field() in regards to km_dokey_event() with the behavior of km_dokey for any dialog,
which should be similar but its not. 
Side note: shouldn't mw_get_field() be MENU aware ? In my sidebar PR I took mw_get_field() made it pass a callback and replaced MENU_EDITOR with MENU_SIDEBAR that's hardly the correct approach. Otoh I don't know mw_get_field should be generic or not.

Come to think of it we could probably alter the order of the elements in haystack to give higher
of lower priority to the keybindings.

Right before your reply I had another reply prepared but I dropped it in fear I was completely off course,
I wanted to pass menutype as flags in km_dokey_event() like: km_dokey_event(MENU_ALIAS|MENU_GENERIC)
but I think I like the Domain thing better.

Anyhow I'm still thinking about all this.
I have to think about this ..  but I might as well start cooking something for 'much further ahead' phase.
Unless you'll change your mind about that and use your magic to do the same think for me too ;)

You do realize if we keep this up we'll end up rewriting NeoMutt from the ground up in one single PR :)

PS. Thanks for taking the time to post really long replies, I figured I could do the same ;)

**Comment by flatcap** [comment]
> Thanks for taking the time to post really long replies

No problem.
There are a lot of moving parts, and the long answers help me understand too :-)

> > The first step is to change km_dokey_event() to take a set of struct MenuFuncOp.
> > If a caller needs Generic functions, then pass, say, { OpAlias, OpGeneric }
> > This means we can eliminate the (mtype == MENU_SIDEBAR) logic and retry_generic().

> So struct MenuFuncOp will act as a haystack, and probably Keymap (MenuType right now) will act as the needle.

Yes

> However I'm unsure about cleaning up MenuFuncOp structures like OpAlias, OpPager, etc.

I have some ideas, but they're not fully-formed yet.

OpGeneric might be the starting point.
It's mostly Menu functions and some Globals.

> there is a somewhat of a duplication in MenuFuncOp and MenuOpSeq (aka DefaultBindings)
> well maybe not duplication, the content of MenuFuncOp and MenuOpSeq are somewhat in sync with each other.

Yes, I've thought the same thing, can we create a combined struct:

```c
{ "previous-entry", OP_PREV_ENTRY, "<up>" },
```

The catch is that some functions have multiple keybindings, e.g.

```c
{ OP_FIRST_ENTRY, "<home>" },
{ OP_FIRST_ENTRY, "="      },
```

> Cleaning MenuFuncOp structures wrt _dispatcher() will probably create a mess wrt to code readability.
> For instance I will have to track code all over the place to find how MENU_ALIAS is actually working:
> km_dokey(MENU_PAGER, {OpPager, OpIndex, OpSidebar, OpGeneric})

Yes and no.
I'm hoping that by splitting things up, the code gets simpler.

I mentioned moving the event loops out of the dialogs...
One of the benefits of this is that the Index and Pager no longer need to know about the Sidebar.
At all!

The user hits a key 'X' and NeoMutt asks the focussed window if **it** can handle it.
If not, NeoMutt traverses all Windows asking if anyone can handle the key.

The Sidebar would be a self-contained unit that observes events and updates its display.

> Come to think of it right now you we're saying the same thing in reply 2 and 3
> (in reply 1 you're telling me to mind my own business - but I ignored that :) )

I'm grateful that you're working on the bigger picture,
but I didn't want you to feel _obliged_ to do so.

> I suppose for now we can throw MenuFuncOp structures as they were and see where this takes us.

Yes, we don't need to alter them yet, but we can start thinking about overlaps.

> I have trouble of mentally isolating the behavior of mw_get_field() in regards to km_dokey_event()
> with the behavior of km_dokey for any dialog, which should be similar but its not.

The main difference is that the Dialogs expect an opcode and mw_get_field() needs both opcodes and raw keys.
(Ignore dlg_select_history(), dlg_select_pattern() and dlg_verify_certificate() -- they're badly written)

> Side note: shouldn't mw_get_field() be MENU aware ?

Hmm... yes

> In my sidebar PR I took mw_get_field() made it pass a callback and replaced MENU_EDITOR with MENU_SIDEBAR that's hardly the correct approach.

It's MENU_SIDEBAR now, but in the future we'll want to pass in other menus.

> Otoh I don't know mw_get_field should be generic or not.

I think we should leave this alone.
Perhaps we can factor out some of the common code with mw_get_field_notify(), but it's probably best to leave that for later

> Come to think of it we could probably alter the order of the elements in haystack to give higher
> of lower priority to the keybindings.

Yes, exactly

> Right before your reply I had another reply prepared but I dropped it in fear I was completely off course,
> I wanted to pass menutype as flags in km_dokey_event() like: km_dokey_event(MENU_ALIAS|MENU_GENERIC)
> but I think I like the Domain thing better.

That's a possibility, but we'd still need logic to tell us which of MENU_ALIAS or MENU_GENERIC comes first.

> I might as well start cooking something for 'much further ahead' phase.

That's great, thanks!

> Unless you'll change your mind about that ;)

I'm still pretty confident the plan will solve a lot of my/our problems.

> You do realize if we keep this up we'll end up rewriting NeoMutt from the ground up in one single PR :)

hehe
It's always hard to find a small piece to fix.

**Comment by flatcap** [comment]
Here's an example of Menu splitting

### OpGeneric

It seems that there are almost no **generic** functions in OpGeneric.
I can see OpMenu, OpMenuTagging and OpGlobals in the future :-)

| Function              | Scope     | OpCode                    |
| :-------------------- | :-------- | :------------------------ |
| `<exit>`              | generic   | `OP_EXIT`                 |
| `<select-entry>`      | generic   | `OP_GENERIC_SELECT_ENTRY` |
|                       |           |                           |
| `<check-stats>`       | global    | `OP_CHECK_STATS`          |
| `<enter-command>`     | global    | `OP_ENTER_COMMAND`        |
| `<redraw-screen>`     | global    | `OP_REDRAW`               |
| `<shell-escape>`      | global    | `OP_SHELL_ESCAPE`         |
| `<show-log-messages>` | global    | `OP_SHOW_LOG_MESSAGES`    |
| `<show-version>`      | global    | `OP_VERSION`              |
| `<what-key>`          | global    | `OP_WHAT_KEY`             |
|                       |           |                           |
| `<bottom-page>`       | menu      | `OP_BOTTOM_PAGE`          |
| `<current-bottom>`    | menu      | `OP_CURRENT_BOTTOM`       |
| `<current-middle>`    | menu      | `OP_CURRENT_MIDDLE`       |
| `<current-top>`       | menu      | `OP_CURRENT_TOP`          |
| `<first-entry>`       | menu      | `OP_FIRST_ENTRY`          |
| `<half-down>`         | menu      | `OP_HALF_DOWN`            |
| `<half-up>`           | menu      | `OP_HALF_UP`              |
| `<help>`              | menu      | `OP_HELP`                 |
| `<jump>`              | menu      | `OP_JUMP`                 |
| `<jump>`              | menu      | `OP_JUMP_1`               |
| `<jump>`              | menu      | `OP_JUMP_2`               |
| `<jump>`              | menu      | `OP_JUMP_3`               |
| `<jump>`              | menu      | `OP_JUMP_4`               |
| `<jump>`              | menu      | `OP_JUMP_5`               |
| `<jump>`              | menu      | `OP_JUMP_6`               |
| `<jump>`              | menu      | `OP_JUMP_7`               |
| `<jump>`              | menu      | `OP_JUMP_8`               |
| `<jump>`              | menu      | `OP_JUMP_9`               |
| `<last-entry>`        | menu      | `OP_LAST_ENTRY`           |
| `<middle-page>`       | menu      | `OP_MIDDLE_PAGE`          |
| `<next-entry>`        | menu      | `OP_NEXT_ENTRY`           |
| `<next-line>`         | menu      | `OP_NEXT_LINE`            |
| `<next-page>`         | menu      | `OP_NEXT_PAGE`            |
| `<previous-entry>`    | menu      | `OP_PREV_ENTRY`           |
| `<previous-line>`     | menu      | `OP_PREV_LINE`            |
| `<previous-page>`     | menu      | `OP_PREV_PAGE`            |
| `<search>`            | menu      | `OP_SEARCH`               |
| `<search-next>`       | menu      | `OP_SEARCH_NEXT`          |
| `<search-opposite>`   | menu      | `OP_SEARCH_OPPOSITE`      |
| `<search-reverse>`    | menu      | `OP_SEARCH_REVERSE`       |
| `<top-page>`          | menu      | `OP_TOP_PAGE`             |
|                       |           |                           |
| `<end-cond>`          | tagging   | `OP_END_COND`             |
| `<tag-entry>`         | tagging   | `OP_TAG`                  |
| `<tag-prefix>`        | tagging   | `OP_TAG_PREFIX`           |
| `<tag-prefix-cond>`   | tagging   | `OP_TAG_PREFIX_COND`      |

**Comment by flatcap** [comment]
NeoMutt Key Bindings: https://gist.github.com/flatcap/fc84519ad7361f6e093eadda039aa0eb

**Comment by lucilanga** [comment]
A few words about my dungeon, err .. that's km_dokey_event().

I was set to do something in the "much further ahead" phase, but it turns out I do not fully understand the km_dokey_event() behavior yet.

Instead Ive been trying to add multiple keymap / menufuncsop, and it gave the chance to debug km_dokey_event() even further  to understand its behavior.

What I came up with was:
```
struct KeyEvent km_dokey_event(struct KeymapSeq *hsmap, struct MenuFuncOps *hsfuncs)
```

hs=haystack

KeymapSeq, MenuFuncOps are structs composed of struct Keymaps, MenuFuncOp.
I iterate them with for and checking != 0 for the end. We could do this more neatly but for now I believe it will suffice.

I call km_dokey_event the following way:
```
struct KeymapSeq hsmap[] = {
    &Keymaps[MENU_ALIAS],
   &Keymaps[MENU_GENERIC],
   0 };
struct MenuFuncOps hsfuncs[] = { OpAlias, OpGeneric, 0 };

return km_dokey_event3(hsmap, hsfuncs).op;
```

Things are working, mostly, there are still things needing to be handled here and there.

Last time we were discussing the possibility of joining KeyBindings and MenuFunc.

> Yes, I've thought the same thing, can we create a combined struct:
> 
> ```c
> { "previous-entry", OP_PREV_ENTRY, "<up>" },
> ```
> 
> The catch is that some functions have multiple keybindings, e.g.
> 
> ```c
> { OP_FIRST_ENTRY, "<home>" },
> { OP_FIRST_ENTRY, "="      },
> ```

I don't necessary see a problem with this.
We traverse the structure with event.ch and stop on the first hit.

I think it will work, but we will mostly invalidate Keymap stuff, I believe that's probably used for macros too.
Perhaps we can make the mountain come to Muhammad by adding MenuFuncOp to the Keymap. Sort of create_bindings but for
MenuFunc taking a hsfuncs structure as input.
Or maybe we could make KeymapSeq take a MenuFuncOp like:


    {  &Keymaps[MENU_ALIAS], OpAlias },
    {  &Keymaps[MENU_GENERIC], OpGeneric }

we would do a single traversal of the structure and see if it's a keybinding (business as usual) or a MenuFuncOp.
But then again we won't do much about removing duplicated parts, perhaps it should not be removed in the first place.
Anyway I'm mostly thinking with my mouth open here.

Better get something debugged.
My dungeon awaits!

Oh, on the same note, how does one debug km_dokey_event() like functions? It's so re-entrant and so used everywhere
that I'll end with too much .neomuttdebug spam to be useful.
Most of the time I've been trying to pass it via mw_get_field() calls and replace mw_get_field() with my crafted mw_get_field_notify() and there I've resorted to fprintf(stderr, "") and that really messes up with my mutt window. It not great but used sparingly I could cope with it.

Lastly, I've had to raise() on if (mtype == 10 && event.op == OP_MAIN_CHANGE_FOLDER) and use gdb to catch the caller, it was OpIndex from dlg_index() obviously, but didn't know that back then.
I'm probably doing this the wrong way, there has to be an easier path!

**Reply by flatcap** [reply]
> how does one debug `km_dokey_event()` like functions?

I tend to start with a lot of `mutt_debug()` logging.
Then I have another window running tail:

```sh
tail -F ~/.neomuttdebug0 | sed 's/.*\]<.> [a-z0-9_]\+() //'
```

GNU's `tail -F` picks up the new file each time NeoMutt renames it.
`sed` strips off the leading junk.

My other debugging tips...

Use the command line to get to the start of your test:

```sh
# Start Neomutt and compose an email
./neomutt -e "push 'mjim@example.com<enter>hello<enter>'"
```

If NeoMutt is running and "at the right place", you can attach a debugger from another window:

```sh
gdb --pid $(pidof neomutt)
```

(or cgdb which is my preferred wrapper)
Then insert breakpoints and continue.

> I've had to raise() on if (mtype == 10 && event.op == OP_MAIN_CHANGE_FOLDER) and use gdb to catch the caller,

gdb lets you set conditional breakpoints

```
b function_name if (param == 42)
```

One final timesaver I like is breakpoint commands.
gdb config:

```
b file.c:234
command
p param1
p param2->member
end
```

When the breakpoint is hit, gdb prints the values of two params automatically.

I'll reply to the rest of your post when I've understood it :-)

**Reply by flatcap** [reply]
> I call km_dokey_event the following way:
> ```c
> struct KeymapSeq hsmap[] = {
>    &Keymaps[MENU_ALIAS],
>    &Keymaps[MENU_GENERIC],
>    0 };
> struct MenuFuncOps hsfuncs[] = { OpAlias, OpGeneric, 0 };
> 
> return km_dokey_event3(hsmap, hsfuncs).op;
> ```

That looks good

> Last time we were discussing the possibility of joining KeyBindings and MenuFunc.

> > The catch is that some functions have multiple keybindings, e.g.

> I don't necessary see a problem with this.

Another problem, that I've just noticed is that some MenuOpSeq have
bindings for other MenuFuncOp, e.g.

```c
const struct MenuOpSeq AliasDefaultBindings[] = { /* map: alias */
  { OP_TAG, "<space>" },
};
```

Also note that the `/* map: alias */` comments on `OpAlias` and `AliasDefaultBindings` are used to generate the docs.
I think we should leave them alone for now.

> Sort of create_bindings but for MenuFunc taking a hsfuncs structure as input.
> we would do a single traversal of the structure and see if it's a keybinding (business as usual) or a MenuFuncOp.

Sorry, you've lost me there.
